manage-data/ingest/transform-enrich/readable-maintainable-ingest-pipelines.md

```diff
@@ -0,0 +1,677 @@
+
+- `ignore_missing`: Useful when you are not sure that the field you want to rename exists.
+- `ignore_failure`: Helps with any failures encountered. For example, the rename processor can only rename to non-existing fields. If you already have the field `abc` and you want to rename `def` to `abc`, the operation will fail.
+
+## Script processor
+
+If no built-in processor can achieve your goal, you may need to use a [script processor](elasticsearch://reference/enrich-processor/script-processor.md) in your ingest pipeline. Be sure to write scripts that are clear, concise, and maintainable.
+
+### Add new fields
+
+All of the above discussed ways to [access fields](#access-fields) and retrieve their values is applicable within the script context. [Null handling](#null-safe-operators) is still an important aspect when accessing the fields.
+
+:::{tip}
+The fields API is the recommended way to add new fields.
+:::
+
+**Fields API**
+We get the following field `cpu.usage` and we want to rename it to `system.cpu.total.norm.pct` which represents a scale from 0-1.0, where 1 is the equivalent of 100%.
+
+```json
+POST _ingest/pipeline/_simulate
+{
+  "docs": [
+    {
+      "_source": {
+        "cpu": {
+          "usage": 90 <1>
+        }
+      }
+    }
+  ],
+  "pipeline": {
+    "processors": [
+      {
+        "script": {
+          "source": """
+            field('system.cpu.total.norm.pct').set($('cpu.usage',0.0)/100.0) <2>
+          """
+        }
+      }
+    ]
+  }
+}
+```
+1. Our field expects 0-1 and not 0-100, we will have to divide by 100 to get the right representation.
+2. The `field` API is exposed as `field(<field name>)`. The `set(<value>)` is responsible for setting the value. Inside we use the `$(<field name>, fallback)` to read the value out of the existing field. Lastly we divide by `100.0`. The `.0` is important, otherwise it will perform an integer only division and return just 0 instead of 0.9.
+
+**No fields API**
+Without the field API this can also be achieved. However there is much more code involved, as we have to ensure that we can walk the full path of `system.cpu.total.norm.pct`.
+
+```json
+{
+  "script": {
+    "source": "
+      if(ctx.system == null){ <1>
+        ctx.system = new HashMap(); <2>
+      }
+      if(ctx.system.cpu == null){
+        ctx.system.cpu = [:]; <3>
+      }
+      if(ctx.system.cpu.total == null){
+        ctx.system.cpu.total = [:];
+      }
+      if(ctx.system.cpu.total.norm == null){
+        ctx.system.cpu.total.norm = [:];
+      }
+      ctx.system.cpu.total.norm.pct = $('cpu.usage', 0.0)/100.0; <4>
+      "
+  }
+}
+```
+1. We need to check whether the objects are null or not and then create them.
+2. We create a new HashMap to store all the objects in it.
+3. Instead of writing `new HashMap()` we can use the shortcut `[:]`.
+4. We perform the same calculation as above and set the value.
```

Comment:
```suggestion
1. Check whether the objects are null or not and then create them.
2. Create a new `HashMap` to store all the objects in it.
3. Instead of writing `new HashMap()`, use the shortcut `[:]`.
4. Perform the same calculation as above and set the value.
```